### DISCUSSION TITLE
My react app can't read image uploaded to supabase storage

### DISCUSSION THREAD

**Discussion by ceno2017** [supabase/supabase#28539] [Questions]
My react app cannot read seem to image I uploaded to supabase.Here is the error I am seeing: {"hint":"rbac_access_denied_matched_policy[api_key_missing]","message":"`apikey` request header or query parameter is either missing or invalid. Double check your Supabase `anon` or `service_role` API key."}

**Comment by MaxTechnics** [comment]
Can you show code on how you load images?

It's possible you don't have improperly configured security policies on your bucket, or are trying to load images from a private bucket expecting it to be a public one.

**Comment by ceno2017** [comment]
`export async function getCabins() {
  let { data, error } = await supabase.from("cabins").select("*");

  if (error) {
    console.error(error);
    throw new Error("Cabins could not be loaded");
  }

  return data;
}


export async function createEditCabin(newCabin, id) {
  const hasImagePath = newCabin.image?.startsWith?.(supabaseUrl);

  const imageName = `${Math.random()}-${newCabin.image.name}`.replaceAll(
    "/",
    ""
  );

  const imagePath = hasImagePath
    ? newCabin.image
    : `${supabaseUrl}/storage/V1/object/public/cabin-images/${imageName}`;

  let query = supabase.from("cabins");

  //A. CREATE/EDIT A NEW CABIN
  if (!id) query = query.insert([{ ...newCabin, image: imagePath }]);

  if (id) query = query.update({ ...newCabin, image: imagePath }).eq("id", id);

  const { data, error } = await query.select().single();

  if (error) {
    //console.log(error);
    throw new Error("Cabin could not be inserted");
  }
  //2. upload image
  const { error: storageError } = await supabase.storage
    .from("cabin-images")
    .upload(imageName, newCabin.image);

  //3. Delete the cabin if there is an error uploading image
  if (storageError) {
    await supabase.from("cabins").delete().eq("id", data.id);
    console.error(storageError);
    throw new Error(
      "Cabin image could not be uploaded and the cabin was not created"
    );
  }
  // console.log(data);
  return data;
}
//3.
import styled from "styled-components";
import { PropTypes } from "prop-types";

import { formatCurrency } from "../../utils/helpers";
import { deleteCabin } from "../../services/apiCabins";
import { useMutation, useQueryClient } from "@tanstack/react-query";
import toast from "react-hot-toast";
import { useState } from "react";
import CreateCabinForm from "./CreateCabinForm";

const TableRow = styled.div`
  display: grid;
  grid-template-columns: 0.6fr 1.8fr 2.2fr 1fr 1fr 1fr;
  column-gap: 2.4rem;
  align-items: center;
  padding: 1.4rem 2.4rem;

  &:not(:last-child) {
    border-bottom: 1px solid var(--color-grey-100);
  }
`;

const Img = styled.img`
  display: block;
  width: 6.4rem;
  aspect-ratio: 3 / 2;
  object-fit: cover;
  object-position: center;
  transform: scale(1.5) translateX(-7px);
`;

const Cabin = styled.div`
  font-size: 1.6rem;
  font-weight: 600;
  color: var(--color-grey-600);
  font-family: "Sono";
`;

const Price = styled.div`
  font-family: "Sono";
  font-weight: 600;
`;

const Discount = styled.div`
  font-family: "Sono";
  font-weight: 500;
  color: var(--color-green-700);
`;

CabinRow.propTypes = {
  cabin: PropTypes.obj,
};

function CabinRow({ cabin }) {
  const [showForm, setShowForm] = useState(false);

  const {
    id: cabinId,
    image,
    name,
    description,
    maxCapacity,
    regularPrice,
    discount,
  } = cabin;

  const queryClient = useQueryClient();
  console.log(image);

  const { isLoading: isDeleting, mutate } = useMutation({
    mutationFn: deleteCabin,
    onSuccess: () => {
      toast.success("Cabin successfully deleted");
      queryClient.invalidateQueries({
        queryKey: ["cabins"],
      });
    },
    onError: (err) => {
      toast.error(err.message);
    },
  });

  return (
    <>
      <TableRow>
        <Img src={image} />
        <Cabin>{name}</Cabin>
        <div>{description}</div>
        <div>Fits up to {maxCapacity}.</div>
        <Price>{formatCurrency(regularPrice)}</Price>
        <Discount>
          {formatCurrency(discount)}
          <div>
            <button onClick={() => setShowForm((show) => !show)}>Edit</button>
            <button onClick={() => mutate(cabinId)} disabled={isDeleting}>
              Delete
            </button>
          </div>
        </Discount>
      </TableRow>
      {showForm && <CreateCabinForm cabinToEdit={cabin} />}
    </>
  );
}

export default CabinRow;

`

**Comment by MaxTechnics** [comment]
I don't see where you're loading your image.
Check in the dashboard if Supabase registers the correct mime type. Sometimes it detects it as plain text.
During upload you can use the `contentType` options to define a proper mime-type

If your bucket is public, make sure it loads the correct path, if it is private, you might need to use `createSignedUrl()` to get a working image url.

https://supabase.com/docs/guides/storage/serving/downloads
https://supabase.com/docs/guides/storage/uploads/standard-uploads?queryGroups=language&language=js#content-type
